api/pkg/controller/kubeletdnat/kubeletdnat_controller.go

```
@@ -0,0 +1,374 @@
+	if !haveJump && len(desiredRules) > 0 {
+		if err := execIptables([]string{"-t", "nat", "-I", "OUTPUT", "-j", ctrl.nodeTranslationChainName}); err != nil {
+			return fmt.Errorf("failed to create jump rule in OUTPUT chain: %v", err)
+		}
+		glog.V(2).Infof("Inserted OUTPUT rule to jump into chain %s.", ctrl.nodeTranslationChainName)
+	}
+
+	// Ensure to masquerade outgoing vpn packets.
+	if !haveMasquerade {
+		if err := execIptables([]string{"-t", "nat", "-I", "POSTROUTING", "-o", ctrl.vpnInterface, "-j", "MASQUERADE"}); err != nil {
+			return fmt.Errorf("failed to create masquerade rule in POSTROUTING chain: %v", err)
+		}
+		glog.V(2).Infof("Inserted POSTROUTING rule to masquerade vpn traffic.")
+	}
+
+	return nil
+}
+
+// getNodeAddresses returns all relevant addresses of a node.
+func getNodeAddresses(node *corev1.Node) []string {
+	addressTypes := []corev1.NodeAddressType{corev1.NodeExternalIP, corev1.NodeInternalIP}
+	addresses := []string{}
+	for _, addressType := range addressTypes {
+		for _, address := range node.Status.Addresses {
+			if address.Type == addressType {
+				addresses = append(addresses, address.Address)
+			}
+
+		}
+	}
+	return addresses
+}
+
+func getInternalNodeAddress(node *corev1.Node) (string, error) {
+	for _, address := range node.Status.Addresses {
+		if address.Type == corev1.NodeInternalIP {
+			return address.Address, nil
+		}
+	}
+	return "", fmt.Errorf("no internal address found; known addresses: %v", node.Status.Addresses)
+}
+
+// getRulesForNode determines the used kubelet address of a node
+// and creates a dnatRule from it.
+func (ctrl *Controller) getRulesForNode(node *corev1.Node) []*dnatRule {
+	rules := []*dnatRule{}
+
+	port := int(node.Status.DaemonEndpoints.KubeletEndpoint.Port)
+	if port <= 0 {
+		port = provider.DefaultKubeletPort
+	}
+
+	internalIP, err := getInternalNodeAddress(node)
+	if err != nil {
+		glog.Errorf("failed to get internal node address: %v", err)
```

Why does the log message say: `failed to ...` but the function returns successfully?
If `getRulesForNode` cannot create the rules for the given input because the given input is not sufficient, `getRulesForNode` should return an error.

The calling function `getDesiredRules` should instead decide if a single failure should make the whole ruleset creation fail

---

Error handling changed as discussed.